### PR TITLE
[SPARK-7790] [SQL] date and decimal conversion for dynamic partition key

### PR DESCRIPTION
**Pull request by adrian-wang**

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -814,4 +814,40 @@ class SQLQuerySuite extends QueryTest {
...
+      // date
+      sql("drop table if exists dynparttest1")
+      sql("create table dynparttest1 (value int) partitioned by (pdate date)")
+      sql(
+        """
+        |insert into table dynparttest1 partition(pdate)
+        | select count(*), cast('2015-05-21' as date) as pdate from src
+      """.stripMargin)
+      checkAnswer(
+      sql(
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Indentation is off, and don't wrap the next line.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 2]
```diff
@@ -814,4 +814,40 @@ class SQLQuerySuite extends QueryTest {
...
+      sql("drop table if exists dynparttest3")
+      sql("create table dynparttest3 (value int) partitioned by (pdec decimal(5, 1))")
+      sql(
+        """
+          |insert into table dynparttest3 partition(pdec)
+          | select count(*), cast('100.12' as decimal(5, 1)) as pdec from src
+        """.stripMargin)
+      checkAnswer(
+        sql(
+          "select * from dynparttest3"),
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Please don't wrap this line.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 3]
```diff
@@ -814,4 +814,40 @@ class SQLQuerySuite extends QueryTest {
...
+        Seq(Row(500, new java.math.BigDecimal("100.1"))))
+    } finally {
+      sql("drop table if exists dynparttest1")
+      sql("drop table if exists dynparttest2")
+      sql("drop table if exists dynparttest3")
+      sql("set hive.exec.dynamic.partition.mode=strict")
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: `SQLTestUtils` provides several utility methods to help eliminating cleanup code like this. For example, you may use the following to simplify this test case:

```scala
withSQLConf("hive.exec.dynamic.partition.mode" -> "strict") {
  withTable("dynparttest1", "dynparttest2", "dynparttest3") {
    // ...
  }
}
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 4]
```diff
@@ -814,4 +814,40 @@ class SQLQuerySuite extends QueryTest {
...
+    try {
+      sql("set hive.exec.dynamic.partition.mode=nonstrict")
+      // date
+      sql("drop table if exists dynparttest1")
+      sql("create table dynparttest1 (value int) partitioned by (pdate date)")
+      sql(
+        """
+        |insert into table dynparttest1 partition(pdate)
+        | select count(*), cast('2015-05-21' as date) as pdate from src
+      """.stripMargin)
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Please fix the indentations here.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveWriterContainers.scala [hunk 5]
```diff
@@ -195,11 +197,21 @@ private[spark] class SparkHiveDynamicPartitionWriterContainer(
...
   }
 
-  override def getLocalFileWriter(row: Row): FileSinkOperator.RecordWriter = {
+  override def getLocalFileWriter(row: Row, schema: StructType): FileSinkOperator.RecordWriter = {
+    def convertToHiveRawString(col: String, value: Any): String = {
+      val raw = String.valueOf(value)
+      schema(col).dataType match {
+        case DateType => DateUtils.toString(raw.toInt)
+        case DecimalType.Unlimited => BigDecimal(raw).toString()
+        case DecimalType.Fixed(_, _) => BigDecimal(raw).toString()
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveWriterContainers.scala]: Seems that these two branches can be merged into this:

```scala
case _: DecimalType => BigDecimal(raw).toString()
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 3 above)
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: BTW, I don't see `dynparttest2` is used anywhere...

**Comment by liancheng** [comment]
@adrian-wang This PR LGTM except for several styling issues. Would you please help rebasing it? Then I can merge this. Thanks!

**Comment by liancheng** [comment]
@adrian-wang For the `sqlContext` method in `SQLTestUtils`, you may override it with `TestHive`. Anyway, that's not a big deal in the scope of this PR.

**Comment by adrian-wang** [comment]
retest this please.

**Comment by yhuai** [comment]
Thanks! I am merging it to branch 1.4 and master.

**Pull request closed without merging**